src/sentry/snuba/metrics.py

```diff
@@ -0,0 +1,329 @@
+    date_range = timedelta(
+        seconds=int(rounding_interval * math.ceil(date_range.total_seconds() / rounding_interval))
+    )
+
+    # TODO: restrictions
+
+    end_ts = int(rounding_interval * math.ceil(to_timestamp(end) / rounding_interval))
+    end = to_datetime(end_ts)
+    # when expanding the rounding interval, we would adjust the end time too far
+    # to the future, in which case the start time would not actually contain our
+    # desired date range. adjust for this by extend the time by another interval.
+    # for example, when "45m" means the range from 08:49:00-09:34:00, our rounding
+    # has to go from 08:00:00 to 10:00:00.
+    if rounding_interval > interval and (end - date_range) > start:
+        date_range += timedelta(seconds=rounding_interval)
+    start = end - date_range
+
+    # snuba <-> sentry has a 5 minute cache for *exact* queries, which these
+    # are because of the way we do our rounding. For that reason we round the end
+    # of "realtime" queries to one minute into the future to get a one-minute cache instead.
+    if end > now:
+        end = to_datetime(ONE_MINUTE * (math.floor(to_timestamp(now) / ONE_MINUTE) + 1))
+
+    return start, end, interval
+
+
+class QueryDefinition:
+    """
+    This is the definition of the query the user wants to execute.
+    This is constructed out of the request params, and also contains a list of
+    `fields` and `groupby` definitions as [`ColumnDefinition`] objects.
+
+
+    Adapted from [`sentry.snuba.sessions_v2`].
+
+    """
+
+    def __init__(self, query_params, allow_minute_resolution=False):
+        self.query = query_params.get("query", "")
+        raw_fields = query_params.getlist("field", [])
+        self.groupby = query_params.getlist("groupBy", [])
+
+        if len(raw_fields) == 0:
+            raise InvalidField('Request is missing a "field"')
+
+        self.fields = {key: parse_field(key) for key in raw_fields}
+
+        start, end, rollup = get_constrained_date_range(query_params, allow_minute_resolution)
+        self.rollup = rollup
+        self.start = start
+        self.end = end
+
+    def get_intervals(self):
+        start = self.start
+        end = self.end
+        delta = timedelta(seconds=self.rollup)
+        while start < end:
+            yield start
+            start += delta
+
+
+class MockDataSource:
+
+    _tags = {
+        "environment": [
+            "production",
+            "staging",
+        ],
+        "release": [  # High cardinality
+            f"{major}.{minor}.{bugfix}"
+            for major in range(3)
+            for minor in range(13)
+            for bugfix in range(4)
+        ],
+        "session.status": [
+            "crashed",
+            "errored",
+            "healthy",
+        ],
+    }
+    _tag_names = sorted(_tags.keys())
+
+    _metrics = {
+        "session": {
+            # "type": "counter",
+            "operations": ["sum"],
```

Comment:
Should this be `count()` or `sum()`? The physical operation is in fact a sum, but since we're dealing with a "counter", I'd rather use `count` here. 

@markstory do you have opinions?

Reply:
It's `sum()` in the [sessions API](https://docs.sentry.io/api/releases/retrieve-release-health-session-statistics/), but that doesn't mean we can't change it. Also, the resulting value will be a float, so it might be weird to get `count(foo) => 123.4`. But I'm open to it.

Reply:
Good point about floats. In that case only `sum` makes sense.
This makes me reconsider if allowing floats for counters is a good idea to begin with :)

Reply:
`sum` makes sense to me. We use sum in sessions and discover in a similar way.